File: 532. K-diff Pairs in an Array.cpp
class Solution {
public:
    /*
    Time: O(nlogn) for sorting
    Space: O(1)

    Two Pointers Approach:
    Start from i = 0th and j =  1st index.

    if their difference is smaller than k, increment j 
    if their difference is greater than k, increment i (if i and j becomes equal, increment j too) 
    if difference = k, increment both and avoid duplicates
    */
    int findPairs(vector<int>& a, int k) { 
      int i = 0, j = 1, ans = 0, n = a.size();
      sort(a.begin(), a.end());

      while(j<n){
        int diff = a[j] - a[i];
        if(diff == k){
          //cout<<i<<" "<<j<<endl;
          ans++;
          j++; i++;
          while(j<n && a[j]==a[j-1])  j++; //avoid duplicates
          while(i<j-1 && a[i]==a[i-1])  i++; //avoid duplicates
        }
        else if(diff < k){
          j++;
        }
        else{
          i++;
          if(i==j)  j++; //condition i != j
        }
      }

      return ans;
    }
};

File: _532_K-diffPairsinanArray.java
// space complexity - o(1)
    //time - o(nlogn + n)
    //Ran on Leetcode successfully : Yes
    // Problem faced  : No
    //Approach : Sort the array and use two sum logic, move both starta nd end pointer from begining of array
      // compare teh differnece of start and end with target and anbsd on the move pointers
      // to remove duplicates , move end till end-1 and end are not same

public int findPairs(int[] nums, int k) {

        Arrays.sort(nums);

        int count  = 0;
        int start = 0;
        int end = 1;

        while(end < nums.length){
            if(start == end){
               end++;
            }
            else{
                if(nums[end] -  nums[start] == k){
                    count++;
                    start++;
                    end++;
                }
                else if(nums[end]  - nums[start] < k){
                    end++;
                }
                else{
                    start++;
                }

                while(end < nums.length && nums[end] == nums[end-1]  ){
                    end++;
                }
            }
        }
        return count;
    }
